<a href="https://colab.research.google.com/github/j-claudinei-f/j-claudinei-f/blob/main/Inducao_e_Recorrencia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Texto auxiliar sobre demonstrações por indução e relações de recorrência**

José Claudinei Ferreira

Universidade Federal de Alfenas

#**Teorema (infinidade de números primos):**

Existem inifnitos números primos.

**Demonstração:** Sabemos que um número natural $n$ é primo quando ele é divisível apenas por 1 e por ele mesmo.

Suponha que o teorema seja falso. Isso implicaria que o conjunto $P$ de todos os números primos seria finito, ou seja, teria $p$ elementos, sendo $p$ um número natural.

Então, poderíamos enumerar esses elementos de $P$ como $$P=\{n_1,\,n_2,\,\ldots,\, n_p\}.$$

Se esse fosse o caso, o número natural $$n_{p+1}=n_1\times n_2\times\cdots\times n_p+1$$ <font color=blue> não seria divisivel por nenhum $n_j\in P$ e, portanto, seria primo.

Isso geraria um problema, porque supomos que existiam apenas $p$ números primos e não $p+1$.

Logo, $P$ não pode ser finito.

**Obs:** Não é sempre verdade que o número $p_{p+1}$ da demonstração é primo. Seria verdade se $P$ fosse finito, o que não é o caso. Mas se você testar alguns casos pode encontrar primos.

In [ ]:
import sympy as sp # para usar a função isprime() que testa se um número é primo.

Testendo alguns casos:

In [ ]:
p=2*3+1
p,sp.isprime(p)

(7, True)

In [ ]:
p=2*3*5+1
p,sp.isprime(p)

(31, True)

In [ ]:
p=2*3*5*7+1
p,sp.isprime(p)

(211, True)

In [ ]:
p=2*3*5*7*11+1
p,sp.isprime(p)

(2311, True)

In [ ]:
p=2*3*5*7*11*13+1
p,sp.isprime(p)

(30031, False)

**Obs:** <font color=blue> Na verdade, na demonstração anterior apresentada usamos um resultado que podemos provar.

#**Teorema:** Um número natural $n$ ou é primo ou é o produto de números primos.

**Demonstração:** <font color=blue> Feito em sala.

**Corolário:** Um número natural $n$ ou é primo ou é o produto de números primos.

**Demonstração:** <font color=blue> Feito em sala.

**Corolário:** Um número natural $n$ ou é primo ou é o produto de $n=n_1\times n_2$, com $n_1$ primo e $1<n_1\leq \sqrt{n}$.

**Demonstração:** <font color=blue> Feito em sala.

Esses resultados nos permite verificar se um número é primo ou determinar seus fatores, como o algoritmo que segue:

In [20]:
import math

def Fatores_primos(n):     # Só funciona para números naturais maiores que 1.
    while n % 2 == 0:      # imprime o número de vezes que 2 divide n.
        print(2)
        n = n // 2         # troca n por n/2 (divisão inteira)

    i=3
    while i*i<=n:          # testando números ímpares.
         while n % i== 0:  # imprime o número de vezes que i divide n.
            print(i)
            n = n // i     # troca n por n/i (divisão inteira)
         i=i+2

    if n > 2:
        print(n) # Se não hove divisão de $n$ ele é primo

**Obs:** Note que, caso prefira (ou possa) usar o cálculo de $\sqrt{n}$ pode trocar o 'while i*i<=n:' por 'for i in range(3,int(math.sqrt(n))+1,2):'.

Teste com $n=2\times3\times5\times7\times11\times13+1=30031$ que já testamos não ser primo.



In [21]:
p=2*3*5*7*11*13+1
Fatores_primos(p)

59
509


Vamos testar com outro número:

In [22]:
Fatores_primos(123456789)

3
3
3607
3803


In [23]:
Fatores_primos(2*3*5*7*11+1)

2311
